### PR TITLE
Add oidc-token-propagation-reactive extension

### LINKED ISSUE #22216: OIDC: cannot use oidc-token-propagation with RESTEasy Reactive
### Description

I wanted to use it to propagate my GitHub access token on a REST call to their API, but it depends on RESTEasy Classic, which clashes with the RESTEasy Reactive I'm using.

### Implementation ideas

We have already created a reactive `OidcClientFilter`  version with @geoand so the same should be done for `oidc-token-propagation` - I'll be OK with the current `oidc-token-propagation` be moved to Quarkiverse if necessary - though may be, since the reactive version will be light, it can coexist in the Quarkus repository.

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #22216

This PR introduces a very light weight `oidc-token-reactive-propagation` extension with a very simple reactive client filter, as an alternative to `oidc-token-propagation`. This new extension does not add yet a more complex code dealing with the token exchange/resign before the propagation which is available in `oidc-token-propagation` as I'd like to see some evidence that such features are used before duplicating them in `oidc-token-reactive-propagation`.

With this extension, and the OIDC code flow where a user is authenticating for ex to GitHub, and an extra remote call is required using the returned access token, one would just attach the filter to the reactive client and it will pick up and propagate the GitHub access token, without the users having to inject that token in the endpoint, manually set it on the client, etc.

`integration-tests/oidc-token-propagation` does test the propagation of the bearer tokens, those coming into the endpoint with `HTTP Authorization`.

So for this extension I've added the test which does show the propagation of the code flow access tokens (returned from Keycloak) - while the docs have been saying it is possible it was not tested.

The test uses `Dev Services For Keycloak`. It uses a `Client` -> `Auth with Code Flow` -> `FrontendResource` -> Propagate token -> `Protected Resource` sequence. OIDC `web-app` configuration is used to secure `FrontendResource` (all the required properties are provided by `Dev Services`) but since `ProtectedResource` receives a propagated bearer token it has to be secured using the OIDC `service` configuration to avoid redirecting, so this is why a custom `TenantResolver` switches to the `bearer` tenant in this case.

Note `500` is returned to the test when a call is mode to the protected resource without the token - because Rest Reactive correctly translates the escaped 401 to 500.

Also, since it is a new extension, I think, if it feels one extension too many, then I'd propose to consider relocating a non-reactive `oidc-token-propagation` extension later, CC @gsmet.

### CODE REVIEW

**Review by FroMage — DISMISSED** [review]
LGTM, except I'm unsure about the rules for the module name, so asking @cescoffier and @gsmet for consistency: should it be the current `quarkus-oidc-token-reactive-propagation` or `quarkus-oidc-token-propagation-reactive`?

**Comment by sberyozkin** [comment]
Hi @FroMage yeah, I was thinking about it but ended up naming it as `quarkus-oidc-token-reactive-propagation` to make it consistent with `quarkus-oidc-client-reactive-filter` :-) where `reactive` is not trailing. I don't mind much if it will have to be renamed

File: extensions/oidc-token-reactive-propagation/runtime/pom.xml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        <artifactId>quarkus-oidc-token-reactive-propagation-parent</artifactId>
+        <groupId>io.quarkus</groupId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-oidc-token-reactive-propagation</artifactId>
+    <name>Quarkus - OpenID Connect Token Reactive Propagation - Runtime</name>
+    <description>Use Reactive RestClient to propagate the incoming Bearer access token or token acquired from Authorization Code Flow as HTTP Authorization Bearer token</description>
```
Comment (geoand) [inline extensions/oidc-token-reactive-propagation/runtime/pom.xml:15]:
```suggestion
    <description>Use Reactive REST Client to propagate the incoming Bearer access token or token acquired from Authorization Code Flow as HTTP Authorization Bearer token</description>
```

**Review by geoand — COMMENTED** [review]
👍🏼

**Comment by FroMage** [comment]
So, let's ask @cescoffier and @gsmet again, and if we don't hear back from them we can just accept this as-is.

**Review by cescoffier — APPROVED** [review]
LGTM!

**Comment by cescoffier** [comment]
|About the module name, let's ask @gsmet, I don't have a strong opinion. The current one does not shock me.

**Comment by sberyozkin** [comment]
@cescoffier Hi Clement

> About the module name, let's ask @gsmet, I don't have a strong opinion. The current one does not shock me.

LOL :-). Sure, lets see what Guillaume's preference is. I nearly went ahead yesterday and renamed it to `oidc-token-propagation-reactive` but then hesitated again, thinking about having `oidc-client-reactive-filter` and `oidc-token-propagation-reactive` name misalignment. I guess the fact I did not follow a convention when I named `oidc-client-reactive-client` does not mean all next OIDC reactive aware extensions should follow the same pattern with the non-trailing `reactive` qualifier, but this pattern is already here with `oidc-client-reactive-client` and I guess it is not too bad. That said, I'm open to renaming if preferred

**Comment by gsmet** [comment]
For me the question is: is it the propagation that is reactive? I think what you are doing here is having a reactive version of token propagation so it makes no sense IMHO to have reactive in the middle.

So I would rather go with `oidc-token-propagation-reactive` as the reactive flavor of OIDC token propagation.

But that is if I have understood the purpose of this extension, which is not a given :).

**Comment by sberyozkin** [comment]
@gsmet 

> So I would rather go with oidc-token-propagation-reactive as the reactive flavor of OIDC token propagation.

+1. As I said I just followed the earlier convention I introduced for `oidc-client-reactive-filter` - but there it is in fact truly reactive; here, the only thing which is happening at the moment is that a reactive filter is used to do a simple propagation - later the token exchange can get added which will be reactive in nature but it is not there now, and I'm fine with not continuing with the same pattern and instead stick to the established convention; thanks

**Comment by sberyozkin** [comment]
@FroMage Steph, I've renamed it to `oidc-token-propagation-reactive` - can you please check if all is good and re-approve if all looks good ?

File: docs/src/main/asciidoc/security-openid-connect-client.adoc [hunk 2]
```diff
@@ -11,7 +11,7 @@ include::./attributes.adoc[]
 This guide explains how to use:
 
  - `quarkus-oidc-client`, `quarkus-oidc-client-reactive-filter` and `quarkus-oidc-client-filter` extensions to acquire and refresh access tokens from OpenID Connect and OAuth 2.0 compliant Authorization Servers such as https://www.keycloak.org[Keycloak]
- - `quarkus-oidc-token-propagation` extension to propagate the current `Bearer` or `Authorization Code Flow` access tokens
+ - `quarkus-oidc-token-propagation` and `quarkus-oidc-token-reactive-propagation` extensions to propagate the current `Bearer` or `Authorization Code Flow` access tokens
```
Comment (FroMage) [inline docs/src/main/asciidoc/security-openid-connect-client.adoc:14]:
```suggestion
 - `quarkus-oidc-token-propagation` and `quarkus-oidc-token-propagation-reactive` extensions to propagate the current `Bearer` or `Authorization Code Flow` access tokens
```

**Review by FroMage — DISMISSED** [review]
Looks like two rename mistakes slipped through, otherwise +1

File: integration-tests/oidc-token-propagation-reactive/pom.xml [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    <parent>
+        <artifactId>quarkus-integration-tests-parent</artifactId>
+        <groupId>io.quarkus</groupId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-integration-test-oidc-token-reactive-propagation</artifactId>
+    <name>Quarkus - Integration Tests - OpenID Connect Token Reactive Propagation</name>
```
Comment (gsmet) [inline integration-tests/oidc-token-propagation-reactive/pom.xml:14]:
```suggestion
    <name>Quarkus - Integration Tests - OpenID Connect Token Propagation Reactive</name>
```

File: integration-tests/oidc-token-propagation-reactive/pom.xml [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    <description>Module that contains OpenID Connect Token Reactive Propagation tests</description>
```
Comment (gsmet) [inline integration-tests/oidc-token-propagation-reactive/pom.xml:15]:
```suggestion
    <description>Module that contains OpenID Connect Token Propagation Reactive tests</description>
```

File: docs/src/main/asciidoc/security-openid-connect-client.adoc [hunk 5]
```diff
@@ -931,6 +931,23 @@ As already noted above, please use `AccessTokenRequestFilter` if you work with K
...
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-oidc-token-propagation-reactive</artifactId>
+</dependency>
+----
+
+`quarkus-oidc-token-propagation-reactive` extension provides `io.quarkus.oidc.token.propagation.reactive.AccessTokenRequestReactiveFilter` which can be used to propagate the current `Bearer` or `Authorization Code Flow` access tokens.
```
Comment (gsmet) [inline docs/src/main/asciidoc/security-openid-connect-client.adoc:947]:
```suggestion
The `quarkus-oidc-token-propagation-reactive` extension provides `io.quarkus.oidc.token.propagation.reactive.AccessTokenRequestReactiveFilter` which can be used to propagate the current `Bearer` or `Authorization Code Flow` access tokens.
```

File: docs/src/main/asciidoc/security-openid-connect-client.adoc [hunk 6]
```diff
@@ -931,6 +931,23 @@ As already noted above, please use `AccessTokenRequestFilter` if you work with K
...
+
+`quarkus-oidc-token-propagation-reactive` extension (as opposed to the non-reactive `quarkus-oidc-token-propagation` extension) does not currently support the exchanging or resigning the tokens before the propagation, however, these features may be added in the future.
```
Comment (gsmet) [inline docs/src/main/asciidoc/security-openid-connect-client.adoc:949]:
```suggestion
The `quarkus-oidc-token-propagation-reactive` extension (as opposed to the non-reactive `quarkus-oidc-token-propagation` extension) does not currently support the exchanging or resigning the tokens before the propagation.
However these features may be added in the future.
```

**Review by gsmet — DISMISSED** [review]
Added a few suggestions.

I suspect it's something we want in 2.7, right?

**Comment by sberyozkin** [comment]
Hi @gsmet Thanks for reviewing it as well, it should be safe to get in 2.7, let me add a backport label, cc @FroMage

**Pull request merged**